authorization/v2/uniformResourceIdentifier.go

```diff
@@ -24,6 +26,79 @@ func NewUriFromRequest(resource resources.Resource, handler gin.HandlerFunc, ctx
 	}
 }
 
+// NewURIFromString parses the string representation of a URI into the UniformResourceIdentifier struct.
+// NewURIFromString expects the string to be of the following form, otherwise ErrInvalidURI is returned.
+// hs:<service_name>:<subsystem>:<version>:<category>:<resource_name>?<allowed_arguments>#<permission_metadata>
+func NewURIFromString(source string) (UniformResourceIdentifier, error) {
+	remainingURI, metadata, err := newURIListFromString(source, "#")
+	if err != nil {
+		return UniformResourceIdentifier{}, errors.Wrap(ErrInvalidURI, err.Error())
```

Comment:
Could you wrap the error returned by `newURIListFromString` with `could not unmarshall metadata`? That way the errors returned by `newURIListFromString` could be generic and you wouldn't need to use messages like `could not unmarshall uri metadata` in a function that is not meant to only unmarshall metadata.

A similar change can be done to line 40 for  the args.

Reply:
Sorry, I don't quite understand what you wanted here. Something like this?

```suggestion
		return UniformResourceIdentifier{}, errors.Wrap(ErrInvalidURI, errors.Wrap(err, "could not unmarshall metadata").Error())
```

Reply:
Yeah, exactly.

Now for example when there's an invalid key value pair in the metadata list we'd get an error: `invalid URI: could not unmarshall metadata: malformed key value pair at index 2`